Update Castle.Windsor.Extensions.DependencyInjection to support .NET8

alkampfergit opened the pull request:
This Pull Request contains also all modification by Alessandro Giorgetti already made in PR #664 related to issue #646 

This Pull Request aim to give Castle.Windsor.Extensions.DependencyInjection full support to .NET 8 new DI model with keyed service. The reason of this Fork was the inability to upgrade ours production code to .NET 8 due to internal error in Castle.Windsor. This problem happens especially with Microsoft Orleans 8 and sometimes code running on Kestrel 8. 

1. Modified code to support KeyedService in .NET 8. All basic test from framework passes, also this pr contains some other tests that idenfy some problem that occurred in our production code
2. Scope management: Orleans 8 changed how it resolves grains so we always got a null Root Scope. This happens because the Root Scope is marked with AsyncLocal, but Orleans uses thread differently so that AsyncLocal is always null. 
3. We introduced a modification in base Castle.Windsor to allow for IHandler to return the current kernel object. While this is somewhat an Hack, it was the only way for us to find the correct root scope when that AsyncLocal is null. This Hack can be avoided if we always use one single instance container, but this is a too restrictive situation.

We actually start using this code in our production code, because without these modification we have as only option to use Orleans 8 to entirely remove Castle.Windsor from our project (An extemely big work) so this code, starting from today, is running in our solution.

Feel free to ask any information or require modification in code if you feel that something is wrong.

Thanks.

AGiorgetti committed:
Refs: #646 Castle.Windsor.Extension.DepencencyInjection removed null

check on scope cache (AsyncLocal can be null on Threads coming from
Threadpool.UnsafeQueueUserWorkItem, having no null check was also the
original behavior)

AGiorgetti committed:
ExtensionContainerRootsScopeAccessor might return a null root scope in

Threadpool.UnsafeQueueUserWorkItem.

AGiorgetti committed:
Added More failing test cases

AGiorgetti committed:
Added Threadpool.Unsafe... failing tests, added comments on critical

files

AGiorgetti committed:
Improved tests and comments

AGiorgetti committed:
Option to map custom NetStatic to Castle Singleton lifestyle.

this improve resolution of singleton object that might not need a root
scope, the windsor container should be enough to resolve singletons

AGiorgetti committed:
Updated Castle.Windsor reference to nuget package to create private

deploy packages

AGiorgetti committed:
RootScope AsyncLocal cache can be null when AspNetCore tries to create

scopes from Threadpool.UnsafeQueueUserWorkItem

AGiorgetti committed:
Improved Tests showing supported, unsupported and memory leaks scenarios

AGiorgetti committed:
Added comments on WindsorScopedServiceProvider dispose

AGiorgetti committed:
Added NuGet package build instructions

alkampfergit committed:
added gitversion

alkampfergit committed:
Update to work with .NET 8 dependency registration.

Major fix was to support keyed registration scenario

alkampfergit committed:
Fixed bug in missing root scope with orleans/kestrel.

The previous handling of root scope is wrong, the code
set root scope in an AsyncLocal variable when WindsorServiceProviderFactoryBase
was first creatd. The problem arise with kestrel or orleans in
.NET 8, they use a Thread Pool that runs code outside AsyncLocal
so it will break resolution.
It was not possible to reproduce locally, but it was reproduced
with production code. A repro for the bug still missing.
Also we can support using a Global root scope only if we use
only ONE CONTAINER in the .NET core DI, because basic structure
does not allow to find the container that is resolving scoped component
thus we cannot determin the right root context.
Still work to do to support multiple container.

alkampfergit committed:
Added logic to support multiple container in .NET 8 DI

Needed to modify basic Castle.Winsor library to support
the ability from IHandler interface to get the current kernel
associated with the handler. This is needed to find the correct
root scope associated with that kernel instance.

AGiorgetti commented on src/Castle.Windsor.Extensions.Hosting/Castle.Windsor.Extensions.Hosting.csproj:
```diff
@@ -23,7 +23,7 @@
 
 	<ItemGroup Condition="'$(TargetFramework)'=='net6.0'">
-    	<PackageReference Include="Microsoft.Extensions.Hosting.Abstractions" Version="6.0.0" />
+    	<PackageReference Include="Microsoft.Extensions.Hosting.Abstractions" Version="8.0.0" />
```
should support proper multitarget, we decided that only net8.0 should have 8.x references (see Castle.Windsor.Extensions.DependencyInjection)

alkampfergit replied:
Fixed

AGiorgetti commented on buildscripts/common.props:
```diff
@@ -18,7 +18,7 @@
 		<FileVersion>$(BuildVersionNoSuffix)</FileVersion>
 		<VersionPrefix>$(BuildVersion)</VersionPrefix>
-		<AssemblyVersion>$(BuildVersionMajor).0.0</AssemblyVersion>
+		<AssemblyVersion>6.0.0.0</AssemblyVersion>
```
revert this file to original

alkampfergit replied:
Reverted

AGiorgetti commented on src/Castle.Windsor.Extensions.DependencyInjection/Scope/ExtensionContainerScopeBase.cs:
```diff
@@ -25,9 +25,14 @@ internal  abstract class ExtensionContainerScopeBase : ILifetimeScope
 
+		private static long _counter;
+
+		public long Counter { get; private set; }
```
was added for debugging purposes, we can get rid of this counter

AGiorgetti commented on src/Castle.Windsor/MicroKernel/ComponentActivator/DefaultComponentActivator.cs:
```diff
@@ -138,6 +138,10 @@ protected virtual object CreateInstance(CreationContext context, ConstructorCand
 		protected object CreateInstanceCore(ConstructorCandidate constructor, object[] arguments, Type implType)
 		{
 			object instance;
+			if (Model.Implementation.FullName.Contains("SiloConnectionFactory")) 
```
remove this code

alkampfergit replied:
done

AGiorgetti reviewed (commented):
There is some cleanup to do!

AlexanderKot commented:
Seems there is related issue
https://github.com/Finbuckle/Finbuckle.MultiTenant/issues/786
Finbuckle multi tenancy + CW as DI container (initially 5.1.2, migration to 6.0.0 do not help)

bxjg1987 commented:
The project uses Windsor 6.0, and after configuring orleans client 8.0, it prompts "This service descriptor is keyed." Your service provider may not support key services

alkampfergit commented:
This pull request introduced keyed registration only in .NET 8, please confirm that your project targets .NET 8.

alkampfergit commented:
Lots of time passed, is there anyone that can comment/reject/accept this Pull Request?

alkampfergit committed:
Changed constants in keyedregistration helper.

TheDutchDev commented:
> Lots of time passed, is there anyone that can comment/reject/accept this Pull Request?

Yeah, what's taking so long? We're running into the same issues here, this PR fixes that..

jklawrence commented:
It would be great to get this support in Castle Windsor.

I did notice that this PR has lots of checks like `#if NET8_0_OR_GREATER`, but I think those could simply be removed. Keyed services are supported in .NET 6.0, 7.0, and even .NET Framework - anywhere Microsoft.Extensions.DependencyInjection is supported.

alkampfergit commented:
From what I know, keyed service provider was introduced in ASP.NET core in .NET 8 and it is not present in previous version. I can be wrong but I was forced to add that directive because code does not compile in .NET 6 (have not tried with 7)

jklawrence commented:
It was released along with .NET 8, but doesn't require .NET 8. 

[Microsoft.Extensions.DependencyInjection 8.0.0](https://www.nuget.org/packages/Microsoft.Extensions.DependencyInjection/#supportedframeworks-body-tab) is supported on multiple frameworks. I've used MS DI with keyed services on .NET Framework just fine. You should just need to update the NuGet package to 8.0.0 in your .NET 6 project and it should compile.

alkampfergit commented:
Perfect, I'll modify the branch as soon as possible. This will force people using castle on .NET 6 to update to version 8 of the DI package but I think that this can be just fine.

alkampfergit committed:
Updated nuget test adapter fixed a resolution bug

The bug happened when you register a service with one NON keyed
component then again with KEYED components. The adapter incorrectly
checked only the first returned service for KEYED and returns null.

Identified after update to Orleans 8.1.0

alkampfergit commented:
If someone used the version locally compiled, I've pushed a fix for a bug that makes impossible to use Microsoft Orleans 8.1.0, the error derived on how the basic Microsoft Dependency injection works with multiple registration of the same NON keyed service. 

Update the test suite, tests are now green, Microsoft.Orleans 8.1.0 works.

alkampfergit commented:
@jklawrence I do not know if any maintainer is checking this PR. I can do the modification you requested, but this will force people using version 8.0.0 of dependency injection even if they use .NET 6 framework. I do not see any problem but I'd like some maintainer have last word on it.

alkampfergit committed:
Fixed IsDefault() usage in DependencyInjectionAdapter

If multiple concrete classes are registered in castle, when
you resolve castle resolves the first one. With Microsoft DI
is the oposite, you want the last registered. The resolution
is now fixed to honor IsDefault() because previous code registered
every component with IsDefault() if it is registered from
the adapter.

alkampfergit commented:
If you are interested, I've updated the PR adding a fix that correctly restore the correct use of IsDefault(). In previous version of the adapter all services registered through the adapter with Microsoft DI services registered componente with the IsDefault() (Due to a different rule in resolution). That actually make not possible to correctly use the IsDefault() anymore. Fixed.

alkampfergit committed:
Fixed (again) resolution rules during resolve.

We added a new concept, an extendede property that
allows the code to understand if the dependency
was registered through the adapter (ServiceCollection)
or directly through the Container.

This allows us to change the resolution rule in case
of multiple services registered with the same name.

jonorossi commented on src/Castle.Windsor/MicroKernel/IHandler.cs:
```diff
@@ -82,5 +82,11 @@ public interface IHandler : ISubDependencyResolver
+		/// Needed to support root scope for multiple container when integrating with .NET 8
+		/// DI engine.
+		/// </summary>
+		IKernel GetKernel();
```
This is a fairly severe breaking change to `IHandler`, need to think about this more.

jonorossi commented on .config/GitVersion.yml:
```diff
@@ -0,0 +1,5 @@
+ignore:
+  sha: []
+merge-message-formats: {}
+mode: ContinuousDeployment
```
Let's keep the pull request to just adding support for .NET 8. The diff is 2600+ LOCs.

jonorossi commented on src/Castle.Windsor.Extensions.DependencyInjection.Tests/Castle.Windsor.Extensions.DependencyInjection.Tests.csproj:
```diff
@@ -1,7 +1,7 @@
 
   <PropertyGroup>
-    <TargetFrameworks>netcoreapp3.1;net6.0</TargetFrameworks>
+    <TargetFrameworks>netcoreapp3.1;net6.0;net8.0</TargetFrameworks>
```
Others have said we can add support for keyed services without adding another compile target, that sounds like a good idea.

Windsor should then support .NET 6 until at least end of support (November 12, 2024) and then upgrade to the next LTS.

jonorossi commented on src/Castle.Windsor.Extensions.DependencyInjection.Tests/WindsorScopedServiceProviderCustomWindsorContainerTests.cs:
```diff
@@ -15,17 +15,70 @@
+
+		//	// Assert
+		//	Assert.IsType<FakeService>(service);
+		//}
```
Commented out code

jonorossi commented on src/Castle.Windsor.Extensions.DependencyInjection.Tests/WindsorScopedServiceProviderCustomWindsorContainerTests.cs:
```diff
@@ -15,17 +15,70 @@
+		//	Assert.IsType<FakeService>(service);
+		//}
+
+#if NET6_0_OR_GREATER
+#endif
```
Empty ifdef

jonorossi commented on src/Castle.Windsor.Extensions.DependencyInjection.Tests/WindsorScopedServiceProviderTests.cs:
```diff
@@ -19,13 +19,36 @@ namespace Castle.Windsor.Extensions.DependencyInjection.Tests
+			return _factory.CreateServiceProvider(container);
+		}
+
+		protected virtual void Dispose(bool disposing)
+		{
+			if (!_disposedValue)
+			{
+				if (disposing)
+				{
+					_factory?.Dispose();
+				}
+
+				_disposedValue = true;
+			}
+		}
+
+		public void Dispose()
+		{
+			// Do not change this code. Put cleanup code in 'Dispose(bool disposing)' method
+			Dispose(disposing: true);
+			GC.SuppressFinalize(this);
 		}
```
Can't this use the test framework tear down?

jonorossi commented on src/Castle.Windsor.Extensions.DependencyInjection/Castle.Windsor.Extensions.DependencyInjection.csproj:
```diff
@@ -16,6 +16,7 @@
 		<TreatWarningsAsErrors>true</TreatWarningsAsErrors>
 		<AssemblyName>Castle.Windsor.Extensions.DependencyInjection</AssemblyName>
 		<RootNamespace>Castle.Windsor.Extensions.DependencyInjection</RootNamespace>
+		<RootNamespace>Castle.Windsor.Extensions.DependencyInjection</RootNamespace>
```
Duplicate line

jonorossi commented on src/Castle.Windsor.Extensions.DependencyInjection/Castle.Windsor.Extensions.DependencyInjection.csproj:
```diff
@@ -30,7 +31,13 @@
-	<ItemGroup>
-		<ProjectReference Include="..\Castle.Windsor\Castle.Windsor.csproj" />
-	</ItemGroup>
+  <ItemGroup Condition="'$(TargetFramework)'=='net8.0'">
+    <PackageReference Include="Microsoft.Extensions.DependencyInjection.Abstractions" Version="8.0.0" />
+    <PackageReference Include="Microsoft.Extensions.Logging.Abstractions" Version="8.0.0" />
+    <PackageReference Include="Microsoft.Extensions.Options" Version="8.0.0" />
+  </ItemGroup>
```
Does Windsor now need these additional `Microsoft.Extensions` libraries?

jonorossi commented on src/Castle.Windsor.Extensions.DependencyInjection/Definitions.cs:
```diff
@@ -0,0 +1,7 @@
+
+namespace System.Runtime.CompilerServices
+{
+	[EditorBrowsable(EditorBrowsableState.Never)]
+	internal static class IsExternalInit { }
```
What's this for?

jonorossi commented on src/Castle.Windsor.Extensions.DependencyInjection/Extensions/WindsorExtensions.cs:
```diff
@@ -49,6 +49,13 @@ public static class WindsorExtensions
 		/// <typeparam name="TService"></typeparam>
 		public static ComponentRegistration<TService> NetStatic<TService>(this LifestyleGroup<TService> lifestyle) where TService : class
 		{
+			// I don't think we need this lifestyle at all, usual Singleton should be good enough;
+			// also we maybe don't need the whole rootscope thing. A normal scope set as current should be enough
+			// otherwise we should revert to static rootscope
```
TODO comments

jonorossi commented on src/Castle.Windsor.Extensions.DependencyInjection/RegistrationAdapter.cs:
```diff
@@ -79,52 +158,125 @@ public static string OriginalComponentName(string uniqueComponentName)
-			return registration.UsingFactoryMethod((kernel) => {
+			return registration.UsingFactoryMethod((kernel) =>
+			{
+				//TODO: We should register a factory in castle that in turns call the implementation factory??
```
TODO comments

jonorossi commented on src/Castle.Windsor.Extensions.DependencyInjection/Scope/ExtensionContainerScope.cs:
```diff
@@ -12,6 +12,8 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+using System;
+
```
Check out the [Castle coding style conventions](https://github.com/castleproject/Home/blob/master/coding-standards.md), they may be a bit strange and different to what is common today, but if we follow them it keeps things more readable and maintainable.

jonorossi commented on src/Castle.Windsor.Extensions.DependencyInjection/Scope/ExtensionContainerScopeCache.cs:
```diff
@@ -20,11 +20,13 @@ namespace Castle.Windsor.Extensions.DependencyInjection.Scope
 		{
-			get => current.Value ?? throw new InvalidOperationException("No scope available");
+			// AysncLocal can be null in some cases (like Threadpool.UnsafeQueueUserWorkItem)
+			get => current.Value; // ?? throw new InvalidOperationException("No scope available. Did you forget to call IServiceScopeFactory.CreateScope()?");
```
Commented out code

jonorossi commented on src/Castle.Windsor.Extensions.DependencyInjection/WindsorScopedServiceProvider.cs:
```diff
@@ -12,67 +12,312 @@
 
 		public void Dispose()
 		{
+			// root scope should be tied to the root IserviceProvider, so
+			// it has to be disposed with the IserviceProvider to which is tied to
 			if (!(scope is ExtensionContainerRootScope)) return;
 			if (disposing) return;
 			disposing = true;
 			var disposableScope = scope as IDisposable;
 			disposableScope?.Dispose();
+			// disping the container here is questionable... what if I want to create another IServiceProvider form the factory?
```
What are these comments for?

jonorossi reviewed (changes requested):
Can you clear out all the other code changes that are not part of this change and look at my comments.

alkampfergit committed:
Cleanup from PR Requests

njnevin commented:
.NET 6 goes out of support tomorrow and Windsor does not yet properly support .NET 8.  What still needs to be done to get this PR over the finish line and get .NET 8 support?

alkampfergit committed:
Remove wrong disposal of container

prochan2 commented:
Any news on this?

alkampfergit commented:
WE are using this version for months in various production systems without hiccups. The problem is that the change are really big, and seems that there is less interest to close this branch. I know that there are areas where the changes are not so clear, I really understand that we stretched some parts, but all modifications were done testing with our systems to ensure that everything runs fine.

We have a fork of this repository and we release version on this branch so actually if you are interested you can build by yourself, test in your system and check if there are any problems. If many persons will use code compiled from this branch maybe maintainers will accepts the pull requests.

zvirja commented:
What is the status of this as of early 2026? It's quite a blocker for us when when updating to .NET 10:
https://github.com/volosoft/castle-windsor-ms-adapter/issues/48

What do people do?
